src/core/thread/mle_router.cpp

```diff
@@ -3135,49 +3135,34 @@ bool MleRouter::IsMinimalChild(uint16_t aRloc16)
 
 void MleRouter::RemoveNeighbor(Neighbor &aNeighbor)
 {
-    switch (mRole)
+    if (mRole == OT_DEVICE_ROLE_CHILD && &aNeighbor == &mParent)
     {
-    case OT_DEVICE_ROLE_DISABLED:
-    case OT_DEVICE_ROLE_DETACHED:
-        break;
+        BecomeDetached();
+    }
 
-    case OT_DEVICE_ROLE_CHILD:
-        if (&aNeighbor == &mParent)
+    if (!IsActiveRouter(aNeighbor.GetRloc16()))
+    {
+        if (aNeighbor.IsStateValidOrRestoring())
         {
-            BecomeDetached();
+            Signal(OT_NEIGHBOR_TABLE_EVENT_CHILD_REMOVED, aNeighbor);
         }
 
-        break;
+        aNeighbor.SetState(Neighbor::kStateInvalid);
+        Get<IndirectSender>().ClearAllMessagesForSleepyChild(static_cast<Child &>(aNeighbor));
+        Get<NetworkData::Leader>().SendServerDataNotification(aNeighbor.GetRloc16());
 
-    case OT_DEVICE_ROLE_ROUTER:
-    case OT_DEVICE_ROLE_LEADER:
-        if (!IsActiveRouter(aNeighbor.GetRloc16()))
+        if (aNeighbor.IsFullThreadDevice())
         {
-            if (aNeighbor.IsStateValidOrRestoring())
-            {
-                Signal(OT_NEIGHBOR_TABLE_EVENT_CHILD_REMOVED, aNeighbor);
-            }
-
-            aNeighbor.SetState(Neighbor::kStateInvalid);
-
-            Get<IndirectSender>().ClearAllMessagesForSleepyChild(static_cast<Child &>(aNeighbor));
-            Get<NetworkData::Leader>().SendServerDataNotification(aNeighbor.GetRloc16());
-
-            if (aNeighbor.IsFullThreadDevice())
-            {
-                // Clear all EID-to-RLOC entries associated with the child.
-                Get<AddressResolver>().Remove(aNeighbor.GetRloc16());
-            }
-
-            RemoveStoredChild(aNeighbor.GetRloc16());
-        }
-        else if (aNeighbor.GetState() == Neighbor::kStateValid)
-        {
-            Signal(OT_NEIGHBOR_TABLE_EVENT_ROUTER_REMOVED, aNeighbor);
-            mRouterTable.RemoveNeighbor(static_cast<Router &>(aNeighbor));
+            // Clear all EID-to-RLOC entries associated with the child.
+            Get<AddressResolver>().Remove(aNeighbor.GetRloc16());
         }
 
-        break;
+        RemoveStoredChild(aNeighbor.GetRloc16());
+    }
+    else if (aNeighbor.GetState() == Neighbor::kStateValid)
+    {
+        Signal(OT_NEIGHBOR_TABLE_EVENT_ROUTER_REMOVED, aNeighbor);
+        mRouterTable.RemoveNeighbor(static_cast<Router &>(aNeighbor));
```

Comment:
if the neighbor to be removed is the parent, we may not want to execute this `else` block?

Reply:
can you explain more about why we may not want to execute the `else` block?

Reply:
first, original code doesn't do this, right?
second, [mParent](https://github.com/openthread/openthread/blob/master/src/core/thread/mle.hpp#L1648) is one variable independent from [mRouterTable](https://github.com/openthread/openthread/blob/master/src/core/thread/mle_router_ftd.hpp#L769)
It may not make sense to have this `else` block when removing parent.

Thoughts?

Reply:
changed accordingly.